rest_log/components/service.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
+# @author Guewen Baconnier <guewen.baconnier@camptocamp.com>
+# @author Simone Orsi <simahawk@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import json
+import traceback
+
+from werkzeug.urls import url_encode, url_join
+
+from odoo import exceptions, registry
+from odoo.http import request
+
+from odoo.addons.component.core import AbstractComponent
+
+
+class RESTServiceDispatchException(Exception):
+
+    rest_json_info = {}
+
+    def __init__(self, message, log_entry_url):
+        super().__init__(message)
+        self.rest_json_info = {"log_entry_url": log_entry_url}
+
+
+class RESTServiceUserErrorException(RESTServiceDispatchException, exceptions.UserError):
+    """User error wrapped exception."""
+
+
+class RESTServiceValidationErrorException(
+    RESTServiceDispatchException, exceptions.ValidationError
+):
+    """Validation error wrapped exception."""
+
+
+class BaseRESTService(AbstractComponent):
+    _inherit = "base.rest.service"
+    # can be overridden to disable logging of requests to DB
+    _log_calls_in_db = True
+
+    def dispatch(self, method_name, *args, params=None):
+        if not self._db_logging_active():
+            return super().dispatch(method_name, *args, params=params)
+        return self._dispatch_with_db_logging(method_name, *args, params=params)
+
+    def _db_logging_active(self):
+        return (
+            request and self._log_calls_in_db and self.env["rest.log"].logging_active()
+        )
+
+    def _dispatch_with_db_logging(self, method_name, *args, params=None):
+        try:
+            result = super().dispatch(method_name, *args, params=params)
+        except exceptions.UserError as orig_exception:
+            self._dispatch_exception(
+                RESTServiceUserErrorException, orig_exception, *args, params=params
+            )
+        except exceptions.ValidationError as orig_exception:
+            self._dispatch_exception(
+                RESTServiceValidationErrorException,
+                orig_exception,
+                *args,
+                params=params,
+            )
+        except Exception as orig_exception:
+            self._dispatch_exception(
+                RESTServiceDispatchException, orig_exception, *args, params=params
+            )
+        log_entry = self._log_call_in_db(
+            self.env, request, *args, params=params, result=result
+        )
+        log_entry_url = self._get_log_entry_url(log_entry)
+        result["log_entry_url"] = log_entry_url
```

Comment:
@simahawk A service can also return a werkzeug response (to return a binary content for exemple). This line will not work if the result of the call to the service method is not a dict

Reply:
IMO in that case you should have binary data wrapped in a dict which may contain metadata.